firmware/components/driver_framebuffer/driver_framebuffer_matrix_drawing.cpp

```diff
@@ -0,0 +1,433 @@
+
+#include "include/driver_framebuffer_internal.h"
+
+#include "sdkconfig.h"
+
+#include <math.h>
+
+#define TAG "fb-drawing"
+
+#ifdef CONFIG_DRIVER_FRAMEBUFFER_ENABLE
+
+#ifdef CONFIG_G_NEW_TRIANGLE
+void driver_framebuffer_triangle(Window* window, float x0, float y0, float x1, float y1, float x2, float y2, uint32_t color)
+{
+	//sort the points such that point 0 is the top and point 2 is the bottom
+	//lower number is higher on screen
+	float temp;
+	if (y1 < y0) { //ensure y1 is under y0
+		//swap points 1 and 0
+		temp = y0;
+		y0 = y1;
+		y1 = temp;
+		temp = x0;
+		x0 = x1;
+		x1 = temp;
+	}
+	if (y2 < y1) { //ensure y2 is under y1
+		//swap points 2 and 1
+		temp = y1;
+		y1 = y2;
+		y2 = temp;
+		temp = x1;
+		x1 = x2;
+		x2 = temp;
+	}
+	if (y2 < y0) { //ensure y2 is under y0
+		//swap points 2 and 0
+		temp = y0;
+		y0 = y2;
+		y2 = temp;
+		temp = x0;
+		x0 = x2;
+		x2 = temp;
+	}
+	if (y1 < y0) { //ensure y1 is under y0 once more
+		//swap points 1 and 0
+		temp = y0;
+		y0 = y1;
+		y1 = temp;
+		temp = x0;
+		x0 = x1;
+		x1 = temp;
+	}
+	
+	// From point 0 to point 1
+	float inc01 = (x1 - x0) / (y1 - y0);
+	float add01 = (x0 / inc01 - y0) * inc01;
+	if (inc01 == 0) add01 = x0;
+	// From point 0 to point 2
+	float inc02 = (x2 - x0) / (y2 - y0);
+	float add02 = (x0 / inc02 - y0) * inc02;
+	if (inc02 == 0) add02 = x0;
+	// From point 1 to point 2
+	float inc12 = (x2 - x1) / (y2 - y1);
+	float add12 = (x1 / inc12 - y1) * inc12;
+	if (inc12 == 0) add12 = x1;
+
+	// if (inc01 > 1000000 || inc02 > 1000000 || inc12 > 1000000 || inc01 < -1000000 || inc02 < -1000000 || inc12 < -1000000) {
```

Comment:
Best remove this

Reply:
The two top lines were not meant to be selected. I was talking about the commented section below them